### PR TITLE
feat: Note Trait System Prototype (Phase 1)

### PR DESCRIPTION
**Pull request by jonathanyeung**
## Phase 1 of Note Trait System Prototype.

Functionality includes
- Basic note trait system scaffolding
- Basic exposure of name and title modification in the note trait interface
- Ability for a user to register a custom note type
- Replacement of Journal Note functionality with a specific note type implementation (as an example)

Docs PR: https://github.com/dendronhq/dendron-site/pull/315

---

# Pull Request Checklist

You can go to [dendron pull requests](https://wiki.dendron.so/notes/adc39825-77a6-46cf-9c49-2642fcb4248e.html) to see full details for items in this checklist.

### Quality Assurance
- [ ] add a [test](https://wiki.dendron.so/notes/cb22bd36-d45a-4ffd-a31e-96c4b39cb7fb.html#writing-tests) for the new feature
- [ ] make sure all the existing [tests](https://wiki.dendron.so/notes/cb22bd36-d45a-4ffd-a31e-96c4b39cb7fb.html#running-all-tests) pass
- [ ] do a spot check by running your feature with our [test workspace](https://wiki.dendron.so/notes/cb22bd36-d45a-4ffd-a31e-96c4b39cb7fb.html#test-workspace)
- [ ] after you submit your pull request, check the output of our [integration test](https://github.com/dendronhq/dendron/actions) and make sure all tests pass
  - NOTE: if you running MacOS or Linux, pay special attention to the Windows output and vice versa if you are developing on Windows

#### Special Cases
- [ ] if your tests changes an existing snapshot, make sure that snapshots are [updated](https://wiki.dendron.so/notes/cb22bd36-d45a-4ffd-a31e-96c4b39cb7fb.html#updating-test-snapshots)
- [ ] if you are adding a new language feature (graphically visible in VS Code/preview/publishing), make sure that it is included in [test-workspace](https://wiki.dendron.so/notes/cb22bd36-d45a-4ffd-a31e-96c4b39cb7fb.html#test-workspace). We use this to manually inspect new changes and for auto regression testing

### Docs
- [ ] Make sure that the PR title follows our [commit style](https://wiki.dendron.so/notes/adc39825-77a6-46cf-9c49-2642fcb4248e.html#commit-style)
- [ ] Please summarize the feature or impact in 1-2 lines in the PR description
- [ ] If your change reflects documentation changes, also submit a PR to [dendron-site](https://github.com/dendronhq/dendron-site) and mention the doc PR link in your current PR

Example PR Description
```markdown
# feat: capitalize all foos

This changes capitalizes all occurrences of `foo` to `Foo`

Docs PR: <URL_TO_DOCS_PR>
```

### First Time PR
- [ ] sign the [CLA](https://en.wikipedia.org/wiki/Contributor_License_Agreement) which will be prompted by our github bot after you submit the PR
- [ ] add your [discord](https://discord.gg/AE3NRw9) alias in the review so that we can give you the [horticulturalist](https://wiki.dendron.so/notes/7c00d606-7b75-4d28-b563-d75f33f8e0d7.html#horticulturalist) badge in our community

### Analytics
- [ ] if you are adding analytics related changes, make sure the [Telemetry](https://wiki.dendron.so/notes/84df871b-9442-42fd-b4c3-0024e35b5f3c.html) docs are updated

### CODE REVIEW

File: packages/plugin-core/src/services/EngineAPIServiceV2.ts [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+  RespRequired,
+  RespV2,
+  SchemaModuleDict,
+  SchemaModuleProps,
+} from "@dendronhq/common-all";
+import { DendronEngineClient, HistoryService } from "@dendronhq/engine-server";
+import _ from "lodash";
+import { getDWorkspace } from "../workspace";
+
+export class EngineAPIServiceV2 implements DEngineClient {
```
Comment (kevinslin) [inline packages/plugin-core/src/services/EngineAPIServiceV2.ts:39]: was this just for testing? any reason why this couldn't be done in `EngineAPIService`?

File: packages/plugin-core/src/services/TypedNoteService.ts [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+export class TypedNoteManager implements TypedNoteService {
+  private cmdRegistar: CommandRegistrar;
+  registeredTypes: NoteType[] = [];
+
+  constructor(context: vscode.ExtensionContext) {
+    this.cmdRegistar = new CommandRegistrar(context);
+  }
+
+  registerType(type: NoteType): RespV2<void> {
+    if (
```
Comment (kevinslin) [inline packages/plugin-core/src/services/TypedNoteService.ts:51]: is there a check for workspace trust before this executes?

**Review by kevinslin — COMMENTED** [review]
Overall, I'm really excited for this and having it roll out. I think we should polish what you have now and use it on our internal note types (eg. journals, tasks, etc) and then iterate (we can do this before exposing it to the public). 

One objection I've had in the past is this idea of multiple types - my mental model of a type is that it refers to one specific thing vs a collection of things. The behavior we're implementing here is more akin to traits in rust, see https://ultrasaurus.com/2020/01/duck-typing-in-rust-trait-vs-type/

How do you feel about renaming this to the `NoteTrait` system?

File: packages/plugin-core/src/services/TypedNoteService.ts (hunk 2 above)
Comment (jonathanyeung) [reply packages/plugin-core/src/services/TypedNoteService.ts]: Added checks.

File: packages/plugin-core/src/services/EngineAPIServiceV2.ts (hunk 1 above)
Comment (jonathanyeung) [reply packages/plugin-core/src/services/EngineAPIServiceV2.ts:40]: Yep, just for prototyping; it's been converged again.

File: packages/common-all/src/types/noteType.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+   * includes the entire hierarchy. Undefined if a Dendron note is not currently
+   * in focus. During onCreate, this will contain the name of the note about to
+   * be created.
+   */
+  currentNoteName?: string;
+
+  /**
+   * Contains any portion of text that is highlighted in the current editor.
+   */
+  selectedText?: string;
```
[the remark this answers is not in the document]
Comment (jonathanyeung) [reply packages/common-all/src/types/noteType.ts]: Will defer on this one to V2.

File: packages/plugin-core/src/services/EngineAPIServiceLegacy.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
+// import {
```
Comment (kevinslin) [inline packages/plugin-core/src/services/EngineAPIServiceLegacy.ts:1]: can we delete this?

File: packages/plugin-core/src/services/NoteTraitService.ts [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+  /**
+   * Contains list of registered Note Traits
+   */
+  readonly registeredTraits: NoteTrait[];
+
+  /**
+   * Register a New Note Trait
+   * @param type
+   */
+  registerTrait(type: NoteTrait): RespV2<void>;
```
Comment (kevinslin) [inline packages/plugin-core/src/services/NoteTraitService.ts:25]: s/type/trait

File: packages/plugin-core/src/services/NoteTraitService.ts [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+  /**
+   * Unregister an Existing Note Trait
+   * @param type
+   */
+  unregisterTrait(type: NoteTrait): RespV2<void>;
```
Comment (kevinslin) [inline packages/plugin-core/src/services/NoteTraitService.ts:31]: s/type/trait

File: packages/plugin-core/src/services/NoteTraitService.ts [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+  /**
+   * Returns the VS Code command that will create a note for the specified when
+   * the command is run.
+   * @param type
+   */
+  getRegisteredCommandForTrait(type: NoteTrait): string | undefined;
```
Comment (kevinslin) [inline packages/plugin-core/src/services/NoteTraitService.ts:38]: s/type/trait

File: packages/plugin-core/src/workspace.ts [hunk 8]
```diff
@@ -383,6 +395,36 @@ export class DendronExtension {
     this.treeViews = {};
     this.webViews = {};
     this.setupViews(context);
+    this._typeRegistrar = new NoteTraitManager(new CommandRegistrar(context));
+
+    // Register any User Defined Note Types
+    const userTypesPath = vscode.workspace.workspaceFile
```
Comment (kevinslin) [inline packages/plugin-core/src/workspace.ts:401]: with native workspaces, the root of workspace might not be the workspace file but where the dendorn.yml is. `getDWorkspace().wsRott` should take care of both cases

Comment (kevinslin) [reply packages/plugin-core/src/workspace.ts:401]: also, `s/userType/userTraits`

File: packages/plugin-core/src/workspace.ts [hunk 9]
```diff
@@ -383,6 +395,36 @@ export class DendronExtension {
...
+      ? path.join(
+          path.dirname(vscode.workspace.workspaceFile?.fsPath),
+          CONSTANTS.DENDRON_USER_NOTE_TRAITS_BASE
+        )
+      : undefined;
+
+    if (userTypesPath && fs.pathExistsSync(userTypesPath)) {
```
Comment (kevinslin) [inline packages/plugin-core/src/workspace.ts:408]: also, `s/userType/userTraits`

File: packages/plugin-core/src/workspace.ts [hunk 10]
```diff
@@ -383,6 +395,36 @@ export class DendronExtension {
...
+      const files = fs.readdirSync(userTypesPath);
+      files.forEach((file) => {
+        if (file.endsWith(".js")) {
+          const typeId = path.basename(file, ".js");
+          this.L.info("Registering User Defined Note Type with ID " + typeId);
```
Comment (kevinslin) [inline packages/plugin-core/src/workspace.ts:413]: s/Type/trait

**Review by kevinslin — CHANGES_REQUESTED** [review]

**Comment by kevinslin** [comment]
As part of this commit, can you also add docs for traits? `eg. dendron.topic.traits`
It can be set to `published: false`. would help with the review and walk through how the end user flow looks like

File: packages/plugin-core/src/commands/RegisterNoteTraitCommand.ts [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+      this.L.error({ error });
+      return { error };
+    }
+    fs.writeFileSync(scriptPath, noteTraitTemplate);
+
+    const newNoteType = new UserDefinedTraitV1(opts.typeId, scriptPath);
+    getExtension().typeRegistrar.registerTrait(newNoteType);
+
+    await VSCodeUtils.openFileInEditor(vscode.Uri.file(scriptPath));
+    return;
```
[the remark this answers is not in the document]
Comment (jonathanyeung) [reply packages/plugin-core/src/commands/RegisterNoteTraitCommand.ts:91]: Without this `return`, it'll actually produce a ts-warning (not all code paths return a val).  Also, The return type of `VSCodeUtils.openFileInEditor` isn't the desired return value of the function.

File: packages/common-all/src/types/foundation.ts [hunk 12]
```diff
@@ -1,6 +1,7 @@
 import { DNoteAnchorPositioned } from "./typesv2";
 import { URI } from "vscode-uri";
 import { DVault } from "./workspace";
+import { NoteTrait } from ".";
```
[the remark this answers is not in the document]
Comment (jonathanyeung) [reply packages/common-all/src/types/foundation.ts:4]: It shouldn't. 

foundation -> NoteTrait
index -> foundation
index -> NoteTrait

NoteTrait doesn't import anything.

File: packages/common-all/src/types/noteTrait.ts [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+
+  /**
+   * Contains any portion of text that is highlighted in the current editor.
+   */
+  selectedText?: string;
+
+  /**
+   * Contains clipboard contents.
+   */
+  clipboard: string;
```
[the remark this answers is not in the document]
Comment (jonathanyeung) [reply packages/common-all/src/types/noteTrait.ts:40]: good point - agree, I think moving to a union type (text|Blob) when the time comes is the right call.

File: packages/common-all/src/types/noteTrait.ts [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+ */
+export type OnCreateContext = {
+  /**
+   * The value of this varies on the context.  During onWillCreate, this
+   * contains the name of the Dendron note that is currently in focus. The name
+   * includes the entire hierarchy. Undefined if a Dendron note is not currently
+   * in focus. During onCreate, this will contain the name of the note about to
+   * be created.
+   */
+  currentNoteName?: string;
```
[the remark this answers is not in the document]
Comment (jonathanyeung) [reply packages/common-all/src/types/noteTrait.ts:30]: There is the potential for a lot of expansion on the interfaces defined in this file.  I'll leave as is for now and let's add more props as we find them necessary.

File: packages/plugin-core/src/commands/CreateNoteWithTraitCommand.ts [hunk 15]
```diff
@@ -0,0 +1,234 @@
...
+        // extraButtons: [MultiSelectBtn.create(false)],
+      };
+      const lc = LookupControllerV3.create(lookupCreateOpts);
+
+      const provider = new NoteLookupProvider("createTypedNote", {
+        allowNewNote: true,
+        forceAsIsPickerValueUsage: true,
+      });
+
+      const onAccepted: OnAcceptHook = async ({
```
[the remark this answers is not in the document]
Comment (jonathanyeung) [reply packages/plugin-core/src/commands/CreateNoteWithTraitCommand.ts:169]: Added behavior for onDidHide, but I'll defer on modifying the interface. I think it needs a little more involved cleanup.  In LookupControllerV3, why does `showQuickPick()` take a Provider as an argument, when there's also a public _provider Member?

File: packages/common-all/src/types/noteType.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (kevinslin) [reply packages/common-all/src/types/noteType.ts]: could you add a v2 task note to capture followup tasks?

File: packages/plugin-core/src/commands/RegisterNoteTraitCommand.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (kevinslin) [reply packages/plugin-core/src/commands/RegisterNoteTraitCommand.ts:91]: got it. can you add a comment then?

**Comment by jonathanyeung** [comment]
> As part of this commit, can you also add docs for traits? `eg. dendron.topic.traits` It can be set to `published: false`. would help with the review and walk through how the end user flow looks like

Added Docs PR: https://github.com/dendronhq/dendron-site/pull/315

**Comment by kevinslin** [comment]
Looks like we have some compiler issues

**Comment by jonathanyeung** [comment]
> Looks like we have some compiler issues

Should be fixed now.

**Pull request merged**